http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -127,6 +127,52 @@ class SamAuthProvider[F[_]: OpenTelemetryMetrics](
     } yield (callerActions, projectCallerActions)
   }
 
+  /**
+   * Get all Sam resource IDs of the given type on which the user is granted permission to `discover`. Returns ids with explicitly
+   * granted roles only; users with workspace `owner` role, for example, can see all runtimes in the workspace, but
+   * `listResourceIds[SamRuntimeResourceId]` is not guaranteed to return all runtimes in that workspace. Works for any
+   * Sam resource type except SharedApp and similar alternative app types; these require a specific Sam resource object
+   * to determine permissions, as these are stored on the individual records.
+   * @param hasOwnerRole whether to return only resources that the user is granted ownership on (has role = `ownerRoleName`)
+   * @param userInfo the current user
+   * @param resourceDefinition the resource archetype (App, Runtime, etc)
+   * @param resourceIdDecoder converts Json -> SamResourceId, needed by SamDao
+   * @param ev the request context
+   * @tparam R SamResourceId
+   * @return List[SamResourceId] of all resources of type R which are discoverable (or owned, if `hasOwnerRole`) by the user
+   */
+  override def listResourceIds[R <: SamResourceId](
+    hasOwnerRole: Boolean,
+    userInfo: UserInfo
+  )(implicit
+    resourceDefinition: SamResource[R],
+    appDefinition: SamResource[AppSamResourceId],
+    resourceIdDecoder: Decoder[R],
+    ev: Ask[F, TraceId]
+  ): F[Set[R]] = {
+    require(resourceDefinition != appDefinition)
```

Comment:
I don't understand why `SamResource[AppSamResourceId]` is passed implicitly, and why this `require` statement exists.

Moreover, we shouldn't use `require` (or `assert`) -- we should be using `F.raiseError()` or equivalent to throw exceptions with reasonable error messages.

Reply:
This was because the assumptions this method works on don't work for the SharedApp SamResource type. That type stores all the metadata about the app type on individual resources, and so you need the context of a single specific ID to read `ownerRoleName` etc. I don't know Scala typeclass syntax well enough to define `R` as "any SamResourceId except AppSamResourceId," and I didn't want to require an additional parameter to do what amounts to compile-time type checking. I could just import the specific value I want to compare (which I think is something like AppSamAction), but it felt safest to feed the expected implicit to the calling context, just like `resourceDefinition`. I still feel wary working with SamAuthProvider's many implicit expectations, so I'm open to alternatives here.

Reply:
tl;dr This method should never called in the implicit context of an AppSamResource. What would you suggest?

Reply:
OK if you just want to throw an exception at runtime, I would do something like this. No need to pass the implicit `appDefinition: SamResource[AppSamResourceId`.

```
for {
      traceId <- ev.ask
      _ <- F.raiseWhen(
        resourceDefinition.resourceType == SamResourceType.App || resourceDefinition.resourceType == SamResourceType.SharedApp
      )(
        AuthProviderException(
          traceId,
          s"[SamAuthProvider.listResourceIds should not be called for App or SharedApp resource type",
          StatusCodes.BadRequest
        )
      )
  …
```

Reply:
I was unsure about that approach because `resourceType` isn't defined at the general level (no ID) for SharedApp SamResource. I think it'd work fine rn to throw if resourceType == App. But I don't know how the code's going to change because the definition is so strange

Reply:
if someone adds a new app type, and doesn't know to update this line, then it won't catch the use of that type. And since SharedApp doesn't extend App I can't identify them as Apps. This isn't happening in the current code and maybe it doesn't matter because hopefully we'll change the client. But it's frustrating.

Reply:
I removed the require but I don't know if removing the implicit is a big enough win to justify the tradeoff in brittleness...

Reply:
We talked this over in DMs and agreed that leaving the implicit in is ok, although the implicit structure in SamAuthProvider is very hard to read